File: 0000-0999/0047.PermutationsII.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans = []
        self.dfs(sorted(nums), [], ans)
        return ans

    def dfs(self, nums, temp, ans):
        if not nums:
            ans.append(temp)
            return
        for i in xrange(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            self.dfs(nums[:i]+nums[i+1:], temp+[nums[i]], ans)

File: 0000-0999/0047.PermutationsII.go
func permuteUnique(nums []int) [][]int {
    sort.Ints(nums)
    n := len(nums)
    perm := make([]int, n)
    used := make([]bool, n)
    ans := make([][]int, 0)
    dfs(nums, perm, used, 0, &ans)
    return ans
}

func dfs(nums, perm []int, used []bool, cur int, ans *[][]int) {
    if cur >= len(nums) {
        tmp := make([]int, len(perm))
        copy(tmp, perm)
        *ans = append(*ans, tmp)
        return
    }
    for i, num := range nums {
        if used[i] || (i > 0 && nums[i] == nums[i-1] && used[i-1]) {
            continue
        }
        perm[cur] = num
        used[i] = true
        dfs(nums, perm, used, cur+1, ans)
        used[i] = false
    }
}